### DISCUSSION TITLE
[Feature] [Golang] Add support for govulncheck in github actions for Go apps.

### DISCUSSION THREAD

**Discussion by rabajaj0509** [apache/skywalking#10483] [Ideas]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Currently, none of our golang applications are checked for vulnerabilities. Since going applications are used as binaries, these are difficult to test for vulnerabilities. All applications that are written in Golang can be scanned using  [govulncheck](https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck) as a part of the GitHub actions.  This highly reduces the possibility of supply chain attacks through dependencies used by our application. 

We would need to write GitHub actions for the following projects:
1) [skywalking-kubernetes-event-exporter](https://github.com/apache/skywalking-kubernetes-event-exporter)
2) [skywalking-cli](https://github.com/apache/skywalking-cli)
3) [skywalking-rover](https://github.com/apache/skywalking-rover)
4) [skywalking-banyandb](https://github.com/apache/skywalking-banyandb)
5) [skywalking-eyes](https://github.com/apache/skywalking-eyes)
6) [skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e)
7) [skywalking-swck](https://github.com/apache/skywalking-swck)
8) [skywalking-satellite](https://github.com/apache/skywalking-satellite)

Related links: 
1) https://go.dev/blog/vuln
2) https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck
3) https://github.com/marketplace/actions/golang-vulncheck

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by Superskyyy** [comment]
Dependabot can only search for vulnerabilities in claimed dependencies in the repository right (We do have dependabot working in the security tab.)? If so then this is a great addition to cover our own code through static analysis.

**Comment by wu-sheng** [comment]
dependabot is working, activated by ASF Infra. We also have sonatype setup for Java.

https://lift.sonatype.com/results/github.com/apache/skywalking/01GK31R4GXW0Y7JZS3BWY047YG?tab=bom-dr


What is the additional value for this?
CVEs should be scan and report in private if possible. 
If we add it to CI, we have to force other contributer to fix things may not relative to them.

**Reply by kezhenxu94** [reply]
Hi. That's not dependency CVE check. It's about our own code's vulnerability check. This is to check the codes in pull requests so we don't merge vulnerable codes. Contributors will only need fix the vulnerable codes they wrote.

**Reply by wu-sheng** [reply]
We keep CVE scans not that public, once you log in the sonatype platform, you can see dependencies CVE warning.

https://lift.sonatype.com/results/github.com/apache/skywalking/01GK31R4GXW0Y7JZS3BWY047YG?tab=dependencies

**Reply by Superskyyy** [reply]
I think sonatype does have some source code vulnerability scanning, just not sure to what extent it covers our needs. Basic things like binding to all interfaces, debug = True can be detected in source code.

**Comment by wu-sheng** [comment]
Moving to discussion.

**Comment by rabajaj0509** [comment]
Let me rephrase my description for better understanding. Few examples of differentiation between Govuln and Dependobot:

1. **Identify security vulnerabilities specific to Golang projects**: Govuln is designed specifically for Golang projects, so it is better equipped to identify security vulnerabilities that are specific to the Golang language and its ecosystem. For example, Govuln can identify vulnerabilities related to Golang's built-in net/http package, which is commonly used in Golang web applications.
2. **Perform static analysis of Golang source code**: Govuln uses static analysis techniques to identify potential security vulnerabilities in Golang source code. This allows it to identify vulnerabilities that may not be easily detected by other tools, such as dynamic analysis tools like Dependobot.
3. **Provide detailed reports on vulnerabilities**: Govuln provides detailed reports on identified vulnerabilities, including information on the nature of the vulnerability, the location in the source code where it was found, and recommendations for how to fix the vulnerability. This level of detail is not typically provided by Dependobot.

In conclusion, while both Govuln and Dependobot are useful tools for enhancing the security of software projects, they serve different purposes and are designed to address different types of security risks. Govuln is specifically designed for Golang projects and uses static analysis techniques to identify vulnerabilities in Golang source code, while Dependobot focuses on managing dependencies and ensuring that software projects are up-to-date and free of known security vulnerabilities.

**Comment by rabajaj0509** [comment]
I think this would be a very useful addition to the ecosystem. What do you think,? if not we can drop the idea. @wu-sheng @Superskyyy @kezhenxu94 

The changes would look similar to this:
```.go
name: Go Vulnerability Checker
on: [push, pull_request]
permissions: read-all
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: ...
      - uses: ...
        with:
          go-version: "1.19.6"
      - run: date
      - run: go install golang.org/x/vuln/cmd/govulncheck@latest && govulncheck ./...
```

**Reply by kezhenxu94** [reply]
Hi I agree it's helpful. And I suggest we can add the check to Makefile (every Go project in skywalking has a Makefile), so we can conveniently run the check locally. And you can run the check in CI like: 

```yaml

name: Go Vulnerability Checker
on: [push, pull_request]
permissions: read-all
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: ...
      - uses: ...
        with:
          go-version: "1.19.6"
      - run: date
      - run: make check-vulnerability 
``'

**Comment by Superskyyy** [comment]
I remember not all actions can freely run in the Apache Org, it requires admin approval right? @kezhenxu94 

Like the paths-filter we use, the one from sentry org works but not the upstream dorny/paths-filter project.

**Reply by kezhenxu94** [reply]
Yes. Is there a GitHub action for that purpose? We can just try it out (perhaps the admin already approved it for other Apache projects), or I can open a ticket to infra team

**Comment by rohan-at-sonatype** [comment]
Hi there, I'm part of the team from Sonatype Lift, which has been running on some of the skywalking repos for a while now.

Sonatype Lift already bundles some go linters and static analysers, namely

- `golangci-lint`
- `gosec`

We're always looking to add new tools to Lift analysis, and `govulncheck` seems right up our alley. This should take us a couple weeks to add (due to timing of some other priorities on our roadmap) but we're happy to do this if it adds value to the team here.

One question - it does look like the `govulncheck` tool is [experimental](https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck#hdr-Limitations), is that something this group is willing to work with?

**Reply by rabajaj0509** [reply]
True, Govulncheck is experimental. However, we use Govulncheck in a controlled and isolated environment (in our case Github actions) to avoid exposing your systems to potential vulnerabilities.

To avoid any mishappenings, we can -- 
1) keep Govulncheck updated with the latest patches and security updates to reduce the risk of exploitation.
2) always verify the results of Govulncheck scans before taking any action based on its recommendations.

Govulncheck can be a helpful tool for identifying vulnerabilities in Golang packages, but it is important to use it with caution and take appropriate measures to reduce any potential risks. This is just my 2 cents, there are really awesome folks in this group that will have a better perspective.

**Comment by rohan-at-sonatype** [comment]
Thanks @rabajaj0509 - My point around `govulncheck` being experimental stems more from the concern that it may be noisy for folks in Skywalking. IF that is not something the group is super concerned about, we can take that back and work thru adding it on the Lift Platform.

The benefit of doing so is that you'll basically offloading the config to us, and findings will show up on your PR.

**Reply by Superskyyy** [reply]
That is awesome, my +1 on this.

**Reply by rohan-at-sonatype** [reply]
Not a problem. We're pulling it into our prioritization. I'll check back here from time to time updating on progress..

**Reply by Superskyyy** [reply]
Thank you Rohan!

**Comment by rohan-at-sonatype** [comment]
@Superskyyy @rabajaj0509 @wu-sheng 

We're looking to have support for `govulncheck` on Sonatype Lift wrapped up by Monday next week. I'm not sure if Skywalking have made some changes to the `config.toml` configuration, that may impact how the change is rolled out.

**Reply by wu-sheng** [reply]
We don't have introduced anything new recently. If that is ready, we could try from [skywalking-cli](https://github.com/apache/skywalking-cli).

**Reply by rabajaj0509** [reply]
Sure, the PR should be pretty straightforward. I will quickly raise a PR for the github-actions in skywalking-cli, if everything goes well, then next week one of the admins can approve and get the code in.

**Reply by rohan-at-sonatype** [reply]
@Superskyyy @rabajaj0509 @wu-sheng 

Support for `govulncheck` is now Live. 

All skywalking repos except for https://github.com/apache/skywalking-rover and https://github.com/apache/skywalking-swck should simply just work.


`skywalking-rover` has a similar `go.mod` under the /test folder (that you might not want to scan, in which case it will Just Work) and `skywalking-swck` will need `.lift.toml` files under the adapter and operator folders

**Reply by kezhenxu94** [reply]
Hi, @rohan-at-sonatype , what do you mean by "should simply just work", we don't even have a `config.toml` file in those projects, we only have the file in `apache/skywalking` project, how would that work in other repositories? And I don't see any difference (any newly added checks or comments, [example](https://github.com/apache/skywalking-eyes/pull/157)) in those repos, any doc about how to add `govulncheck` in the Go repos?

**Reply by Superskyyy** [reply]
@rohan-at-sonatype Friendly ping :) Any suggestions?

**Reply by rohan-at-sonatype** [reply]
Hey all, sorry for not responding. Totally missed the ping.

@kezhenxu94 it looks like Sonatype Lift is not installed for `apache/skywalking-eyes` cc @wu-sheng

**Reply by Superskyyy** [reply]
> Hey all, sorry for not responding. Totally missed the ping.
> 
> @kezhenxu94 it looks like Sonatype Lift is not installed for `apache/skywalking-eyes` cc @wu-sheng

Thanks for the reply! So basically, it means - once Sonatype is enabled in such repos, `govulncheck` will automatically work like the current enabled repos in a form like comments within PRs, right?

**Reply by rohan-at-sonatype** [reply]
That is correct.
